Core/include/System.h

```diff
@@ -0,0 +1,54 @@
+#ifndef SYSTEM_H
+#define SYSTEM_H
+
+#include <mutex>
+#include <thread>
+#include <chrono>
+#include <iostream>
+#include <atomic> // Include atomic for thread-safe boolean
+
+class System {
+private:
+    std::atomic<bool> isActive;
+    std::thread updateThread;
+    std::mutex mutex;
+
+public:
+    System() : isActive(false) {} // Initialize isActive to false
+    ~System() {
+        Kill();
+    }
+
+    virtual void Initialize(double deltaTimeSecond) {
+        std::lock_guard<std::mutex> lock(mutex);
+        if (!isActive) {
+            isActive = true;
+            updateThread = std::thread(&System::Update, this, deltaTimeSecond);
+        }
+    }
+
+    void Update(double deltaTimeSecond) {
+        while (IsActive()) {
+            std::cout << "Running " << this->GetName() << std::endl;
+            std::this_thread::sleep_for(std::chrono::milliseconds(static_cast<int>(deltaTimeSecond * 1000)));
```

Comment:
`sleep_for` is necessary, since we are using an event-driven design
This would also be important as it's basically serving as our refresh-rate across any of the systems.

Reply:
Doesn't the OS handle time slicing for each thread? For FixedUpdate we would have a set interval, but for Update we want the system to finish the frame as fast as possible. Here we are sleeping for (deltaTimeSecond * 1000)ms which is a huge amount of time for a game engine. 


 If we have a 60FPS target, that means we get 0.016s per frame, and here we are feeding that into this function and sleeping for that time. That means after the first frame, if it even took 0.016s, we are now sleeping for that entire time just for one thread, which means this frame is going to take even longer for the next deltaTimeSecond, this is going to grow until infinity and then the engine is going to stop waiting for infinite seconds for one thread. That's if im assuming you're usingn deltaTimeSecond in the same way unity does.

If you are using deltaTimeSecond as a fixed interval to "wait" for, this is incorrect to use in Update and should be used in FixedDeltaTime.

Reply:
*Correction i mean fixed update, and FixedUpdate should be using fixedDeltaTime or some fixed time. 

Another thing is even though we are waiting for a set time in FixedUpdate, we stil want whatever is happening inside FixedUpdate to finish as fast as possible, we shouldnt be sleeping threads unless there is some blocking operation, but in that case we should be using condition variables and manual blocking so the OS picks up the thread as fast as possible.

Reply:
I would personally like this functionality for Physics' update loop to run at a fixed time-step, so we can enforce a 60hz, 120hz, whatever update for it (physics being fixed-time-stepped is *very important*). Of course, other systems like rendering would probably run as fast as it possibly can - in that case, do they just override this update method?

>If you are using deltaTimeSecond as a fixed interval to "wait" for, this is incorrect to use in Update and should be used in FixedDeltaTime.

I believe they're doing this. It's called an Update(deltaTime) which evokes the concept of a Unity component's update method but its actually just running an infinite (or rather until the isActive atomic bool is set false) while loop and sleeping at the end of each loop for "deltaTime" time to enforce a fixed-time-step.

Reply:
>Another thing is even though we are waiting for a set time in FixedUpdate, we stil want whatever is happening inside FixedUpdate to finish as fast as possible, we shouldnt be sleeping threads unless there is some blocking operation, but in that case we should be using condition variables and manual blocking so the OS picks up the thread as fast as possible.

Agreed. I'm a fan of keeping a time buffer and decrementing it inside of the while loop for a fixed-time-step system instead of sleeping, like below pseudo code:

```
float buffer += delta_time;
while (buffer >= FIXED_TIME_INTERVAL) {
   fixedUpdateShit();
   timeBuffer -= FIXED_TIME_INTERVAL;
}
```

where FIXED_TIME_INTERVAL is the tick-rate of our fixed-time-stepped system. This way you don't have to deal with the overhead of OS time-based-sleeps. There's an argument about there being a slight trade-off of wasted CPU cycles (in a similar vein to that of busy waiting) with this solution but OS sleep methods in the ms time quantums we're working in (like 16ms or 8ms e.g.) will lead to "dropped" frames as you're *still at the whim of the OS scheduler*. It _will_ add milliseconds onto it, it only guarantees a _minimum_ sleep time.

Reply:
If the goal of `sleep for` is to act as filler time to meet a defined interval, then it's wrong. With the current implementation, the thread will sleep for the entire length of the defined interval **even if the code before it passes it**. 
![image](https://github.com/user-attachments/assets/2443ff1f-56d1-46eb-bb70-61d8f3ff75e6)
The bottom chart is what we want to happen in a FixedUpdate, but the top one is whats happening in the code.

Regardless, `Update` should not be running at a fixed interval either way. 
> It's basically serving as our refresh-rate.
The only thing that should be linked to the refresh rate is the Renderer. Refresh rate is the frequency a computer monitor can update. There's no reason for anything other than rendering to be linked directly to the refresh rate.

Reply:
> If the goal of `sleep for` is to act as filler time to meet a defined interval, then it's wrong. With the current implementation, the thread will sleep for the entire length of the defined interval **even if the code before it passes it**. ![image](https://private-user-images.githubusercontent.com/21136883/371920032-2443ff1f-56d1-46eb-bb70-61d8f3ff75e6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc2NjIzMjEsIm5iZiI6MTcyNzY2MjAyMSwicGF0aCI6Ii8yMTEzNjg4My8zNzE5MjAwMzItMjQ0M2ZmMWYtNTZkMS00NmViLWJiNzAtNjFkOGYzZmY3NWU2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTMwVDAyMDcwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNjNzRmNTE4NTEzMWFjYTBlODU3MWI4NGEzNTVhMGZkNjkyNTJmOGE1YThkYTcwMWJlNTM4OTY1YjJhOTg1M2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.NZvMhA2jZ5hlvJZ_fg9gJHMlIFG1Z5gDQDZVuFO2vmg) The bottom chart is what we want to happen in a FixedUpdate, but the top one is whats happening in the code.
> 
> Regardless, `Update` should not be running at a fixed interval either way.
> 
> > It's basically serving as our refresh-rate.
> 
> The only thing that should be linked to the refresh rate is the Renderer. Refresh rate is the frequency a computer monitor can update. There's no reason for anything other than rendering to be linked directly to the refresh rate.

In that case, assuming an updated version of the current system that:

1. gets system time before running logic (lets call it currTime)
2. gets time after running logic (lets call it newTime)
3. then sleeps for 16ms - (newTime - prevTime) -- assuming that result is >0.0

We will have smaller time quantums and run into "dropped" frames more often. Sleep = bu hao.